cmd/vet/defaulthttp.go

```diff
@@ -0,0 +1,94 @@
+// This file contains the defaulthttp checker.
+
+package main
+
+import (
+	"go/ast"
+	"go/types"
+	"path/filepath"
+	"strings"
+)
+
+func init() {
+	register("defaulthttp",
+		"check use of the http.DefaultClient in Chain packages",
+		checkHttpCall,
+		callExpr)
+}
+
+var prohibitedFunctions = map[string]bool{
+	"net/http.Get":      true,
+	"net/http.Head":     true,
+	"net/http.Post":     true,
+	"net/http.PostForm": true,
+}
+
+var prohibitedMethods = map[string]bool{
+	"(*net/http.Client).Do":       true,
+	"(*net/http.Client).Get":      true,
+	"(*net/http.Client).Head":     true,
+	"(*net/http.Client).Post":     true,
+	"(*net/http.Client).PostForm": true,
+}
+
+var checkCmds = map[string]bool{
+	"cored":   true,
+	"corectl": true,
+	"signerd": true,
+}
+
+// checkHttpCall checks that the chain packages do not use the
+// http DefaultClient which does not include TLS settings.
+func checkHttpCall(f *File, node ast.Node) {
+	// Skip test files.
+	if strings.HasSuffix(f.name, "_test.go") {
+		return
+	}
+	// Skip commands besides the ones we explicitly want to check.
+	if f.pkg.path == "main" && !checkCmds[filepath.Dir(f.name)] {
+		return
+	}
+
+	call, ok := node.(*ast.CallExpr)
+	if !ok {
+		return
+	}
+	sel, ok := call.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return
+	}
+	fun, ok := f.pkg.uses[sel.Sel].(*types.Func)
+	if !ok {
+		return
+	}
+	// Check for uses of the http convenience functions that implicitly use
+	// the net/http.DefaultClient.
+	if prohibitedFunctions[fun.FullName()] {
+		f.Badf(call.Pos(), "use of %s", fun.FullName())
+	}
+
+	// It might still be an invocation of a method with the
+	// http.DefaultClient as a receiver.
+	if !prohibitedMethods[fun.FullName()] {
+		return
+	}
+	clientSel, ok := sel.X.(*ast.SelectorExpr)
+	if !ok {
+		return
+	}
+	obj, ok := f.pkg.uses[clientSel.Sel]
+	if !ok {
+		return
+	}
+	if !isPackageLevel(obj) {
+		return
+	}
+	if obj.Pkg().Path() != "net/http" || obj.Name() != "DefaultClient" {
```

Comment:
I think this would not find cases where we've assigned DefaultClient to another variable, right?

```go
x := http.DefaultClient
x.Post(...)
```

That is probably fine to start, but there's one way in which this pattern is likely to come up: when we're using DefaultClient as a default value for an HTTP client, like in https://github.com/chain/chain/blob/6d40c894/core/rpc/rpc.go#L126-L130:

```go
client := c.Client
if client == nil {
	client = http.DefaultClient
}
resp, err := client.Do(req.WithContext(ctx))
```

One way to thoroughly check for this would be to use https://godoc.org/golang.org/x/tools/go/pointer

```go
config := &pointer.Config{
	// ...
}
config.AddQuery(/* http.DefaultClient */)
config.AddQuery(/* obj */)
result, err := pointer.Analyze(config)
objPtr := result.Queries[/* obj */]
defPtr := result.Queries[/* http.DefaultClient */]
if objPtr.MayAlias(defPtr) {
	fail()
}
```

However, I think doing the analysis separately like this for every call site would be too slow, and even if we did one big batch analysis it still might be too slow.